### DISCUSSION TITLE
The essence of Properties parameter on the @Index decorator

### DISCUSSION THREAD

**Discussion by Lishenga** [mikro-orm/mikro-orm#1689] [Q&A] [answered]
Hi, I would appreciate kindly if someone could explain to me the essence or the usage of the properties parameter on the **@Index** decorator.

```ts
@Index({ properties: ['name', 'caption'], type: 'text' })
```

The properties parameter is provided or needed on which instances, is it a must i provide the array of properties and what is its usage?

Thank you in advance.

**Comment by B4nan** [comment, marked as answer]
It can be a single string too, and they are required only at the entity level - it you use the decorator on a property, it will be inferred automatically.

**Reply by Lishenga** [reply]
When you supply the decorator at entity level, what is the essence of the properties parameter?

**Reply by B4nan** [reply]
Essence? I don't follow.

You want to create index over one or more columns, those columns are represented as properties on the entity. And that is what you should put there - property names of your entity.

**Reply by Lishenga** [reply]
so the properties parameter takes an array of column names of the specific entity class and uses them to create the name of the index?

**Reply by B4nan** [reply]
Nope, not column names, property names. That is why it is called `properties` and not `columns`. And nope, it is not about index name, it is about what columns the index will be build upon.

**Reply by Lishenga** [reply]
ooooh okay thank you have clarified on it. So if i have an entity like this 
```ts
@Index({ properties: ['hashChainLength', 'hashMaxJump'], type: 'text' })
export class CodeSystem extends RootEntity {

    [EntityRepositoryType]?: CodeSystemRepository;

    // The length of hashChain.  Used this to initialize the codeGenrator
    // @Field(type => Int)
    @Field(type => Int, { defaultValue: 100000, description: "Default code systems is ACP1, with hashChainLength = 100000." })
    @Property({ default: 100000 })
    public hashChainLength: number;
    // public hashChainLength: number = 100000;

    // This is a algorithm specific value
    // Used for determining if a hashJump is out of bound
    // @Field(type => Int)
    @Field(type => Int, { defaultValue: 1096, description: "Default code systems is ACP1, with hashMaxJump = 1096." })
    @Property({ default: 1096 })
    public hashMaxJump: number;
    // hashMaxJump: number = 1096;
}
```

the index will be built on the properties hashChainLength and hashMaxJump while the type of index will be _text_ ?

**Reply by B4nan** [reply]
Yes. Btw you are using mongo right? If so, the default in `@Property({ default: 1096 })` won't do anything - that is for SQL drivers and serves only at the schema level (native column defaults). You should use the commented way to have default value on runtime instead.

You can check the decorator reference, where SQL only options are marked:

https://mikro-orm.io/docs/decorators#property
